packages/config-loader/src/schema/compile.ts

```diff
@@ -101,17 +108,38 @@ export function compileConfigSchemas(
 
   const merged = mergeConfigSchemas(schemas.map(_ => _.value));
 
-  if (options?.noUndeclaredProperties) {
-    traverse(merged, (schema: SchemaObject) => {
-      /**
-       * The `additionalProperties` key can only be applied to `type: object` in the JSON
-       *  schema.
-       */
-      if (schema?.type === 'object') {
-        schema.additionalProperties ||= false;
+  traverse(
+    merged,
+    (
+      schema: SchemaObject,
+      _: string,
+      _1?: any,
+      _2?: any,
+      _3?: any,
+      parentSchema?: SchemaObject,
+    ) => {
+      // Inherit parent deepVisibility if we don't define one ourselves.
+      if (parentSchema?.deepVisibility) {
+        schema.deepVisibility ??= parentSchema?.deepVisibility;
       }
-    });
-  }
+
+      // This runs before we compile the AJV so visibilityByDataPath has the
+      //  correct data.
+      if (schema.deepVisibility && parentSchema) {
+        schema.visibility ??= schema.deepVisibility;
```

Comment:
Same as above, thinking that we'd want to reject this if `schema.visibility` is `frontend` and `schema.deepVisibility` is `secret`, or the other way around

Reply:
How do we want to handle `backend`? Similar to my other comment, could we declare the default visibility as `backend` on the schema specifically and then use it here? Wouldn't have to happen as part of this PR but food for thought.

Reply:
Although a tricky thing to handle is e.g. the heirarchy deep frontend > deep backend > deep secret. It may be that we want to somehow deeply mark config once it's been marked with either deep frontend or deep secret, and make sure we never flip over completely

Reply:
Yup, was just thinking of that when I saw your comment. That one is likely going to be tricky but important to get right.

Reply:
I tried to give it a go, can you take another look?